### DISCUSSION TITLE
Фильтр для нескольких пользователей одной точки

### DISCUSSION THREAD

**Discussion by horyachka** [totumonline/totum-ru-issues-and-discussions#5] [Как это сделать?] [answered]
Добрый день.
Подскажите пожалуйста как правильно создать префильтр пользователя при таком условии:
1. Пользователи должны видеть только свои строки.
2. Если одна точка имеет двух и более пользователей, то эти пользователи должны видеть строки других пользователей только этой точки.

Пример: Торговый Центр возвращает Молокозаводу поддоны и заполняет заявку. От этого ТЦ данные внесли два пользователя
![image](https://user-images.githubusercontent.com/91678200/135467853-1457b895-ccee-45c8-ba2c-fc471075c6b4.png)

Если каждая точка имеет одного пользователя, тогда префильтр такой:

Name поля в строчной части:
multiselekt
Код:
= : if (condition: $#nr = $roles; then: $#nu; else: "*ALL*")
roles: listCreate(item: 12; item: 15)

Код селекта:
= : selectRowListForSelect(table: 'users'; field: 'fio'; where: 'roles' = $roles)
roles: listCreate(item: 12; item: 15)

При таком префильтре пользователи ТЦ-1-1 и ТЦ-1-2 видят только свои строки по одной, а Молокозавод - две.

Подскажите пожалуйста как написать скрипт, что бы пользователи ТЦ-1-1 и ТЦ-1-2, объединенные в ТЦ-1 таблицы users, могли видеть строки друг друга, а остальные участники только свои строки? (Префильтр должен обеспечить возможность заполнить заявку несколькими пользователями одной точки на выдачу и на прием поддонов. И, скорей всего, в заявке на выдачу в полях Отправитель и Получатель нужно брать данные из поля  name_tochki  таблицы users).

Спасибо!

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Добрый день. Спасибо за вопрос — это достаточно распространенная ситуация, когда надо сделать префильтр по группе, а не по одному значению.

Делается это через дополнительное техническое поле в префильтре.

Попробую объяснить c небольшими упрощениями в ваше задаче.

У вас в строчной части указаны два пользователя — `отправитель` и `получатель`. Ситуацию с `получателем` я приму как стандартную — поле префильтра привязано к полю в строчной части. Когда мы выберем `получателя` в префильтре, мы будем видеть только относящиеся к нему строки.

![получатель](https://totum.totum.online/fls/365_390_file.png)

С `отправителем` история отличается, тк несколько разных `отправителей` могут быть из одной `компании` и должны видеть строки друг-друга, но не видеть остальных.

![отправители](https://totum.totum.online/fls/365_391_file.png)

Для этого, мы префильтр `отправителей` **не привязываем** к полю в строчной части! Вместо этого мы создаем еще одно техническое поле в префильтре. Ставим ему тип Данные и делаем видимым только Создателю. Для большей безопасности рекомендуется сделать его нередактируемым.

Именно этот технический префильтр мы привязываем к `отправителям` в строчной части.

Теперь нам нужно рассчитать в нем значение всех пользователей которые относятся к компании `отправителя` выбранного в префильтре:

```totum
// Код в поле fl_teh_prefiltr 

= : if(condition: #fl_polzovatel_1 = ""; then: $#lc; else: $calc)

calc: selectList(table: 'tablica_s_polzovatelyami'; field: 'id'; where: 'kompaniya' = $company)
	company: select(table: 'tablica_s_polzovatelyami'; field: 'kompaniya'; where: 'id' = #fl_polzovatel_1)
```

Таким образом, когда мы выберем `отправителя = 1`, фильтрация бедет по пользователям `[1,2]` тк они входят в одну `компанию` (на скрине нихе префильтр еще не применен, что бы было нагляднее):

![тех префильтр](https://totum.totum.online/fls/365_392_file.png)
